Mathlib/LinearAlgebra/SModEq.lean

```diff
@@ -0,0 +1,121 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+
+! This file was ported from Lean 3 source module linear_algebra.smodeq
+! leanprover-community/mathlib commit 146d3d1fa59c091fedaad8a4afa09d6802886d24
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Data.Polynomial.Eval
+import Mathlib.RingTheory.Ideal.Quotient
+
+/-!
+# modular equivalence for submodule
+-/
+
+
+open Submodule
+
+open Polynomial
+
+variable {R : Type _} [Ring R]
+
+variable {M : Type _} [AddCommGroup M] [Module R M] (U U₁ U₂ : Submodule R M)
+
+variable {x x₁ x₂ y y₁ y₂ z z₁ z₂ : M}
+
+variable {N : Type _} [AddCommGroup N] [Module R N] (V V₁ V₂ : Submodule R N)
+
+/-- A predicate saying two elements of a module are equivalent modulo a submodule. -/
+def SModEq (x y : M) : Prop :=
+  (Submodule.Quotient.mk x : M ⧸ U) = Submodule.Quotient.mk y
+#align smodeq SModEq
+
+-- mathport name: «expr ≡ [SMOD ]»
+notation:50 x " ≡ " y " [SMOD " N "]" => SModEq N x y
+
+variable {U U₁ U₂}
+
+protected theorem SModEq.def :
+    x ≡ y [SMOD U] ↔ (Submodule.Quotient.mk x : M ⧸ U) = Submodule.Quotient.mk y :=
+  Iff.rfl
+#align smodeq.def SModEq.def
+
+namespace SModEq
+
+theorem sub_mem : x ≡ y [SMOD U] ↔ x - y ∈ U := by rw [SModEq.def, Submodule.Quotient.eq]
+#align smodeq.sub_mem SModEq.sub_mem
+
+@[simp]
+theorem top : x ≡ y [SMOD (⊤ : Submodule R M)] :=
+  (Submodule.Quotient.eq ⊤).2 mem_top
+#align smodeq.top SModEq.top
+
+@[simp]
+theorem bot : x ≡ y [SMOD (⊥ : Submodule R M)] ↔ x = y := by
+  rw [SModEq.def, Submodule.Quotient.eq, mem_bot, sub_eq_zero]
+#align smodeq.bot SModEq.bot
+
+@[mono]
+theorem mono (HU : U₁ ≤ U₂) (hxy : x ≡ y [SMOD U₁]) : x ≡ y [SMOD U₂] :=
+  (Submodule.Quotient.eq U₂).2 <| HU <| (Submodule.Quotient.eq U₁).1 hxy
+#align smodeq.mono SModEq.mono
+
+@[refl]
+protected theorem refl (x : M) : x ≡ x [SMOD U] :=
+  @rfl _ _
+#align smodeq.refl SModEq.refl
+
+protected theorem rfl : x ≡ x [SMOD U] :=
+  SModEq.refl _
+#align smodeq.rfl SModEq.rfl
+
+instance : IsRefl _ (SModEq U) :=
+  ⟨SModEq.refl⟩
+
+@[symm]
+nonrec theorem symm (hxy : x ≡ y [SMOD U]) : y ≡ x [SMOD U] :=
+  hxy.symm
+#align smodeq.symm SModEq.symm
+
+@[trans]
+nonrec theorem trans (hxy : x ≡ y [SMOD U]) (hyz : y ≡ z [SMOD U]) : x ≡ z [SMOD U] :=
+  hxy.trans hyz
+#align smodeq.trans SModEq.trans
+
+theorem add (hxy₁ : x₁ ≡ y₁ [SMOD U]) (hxy₂ : x₂ ≡ y₂ [SMOD U]) : x₁ + x₂ ≡ y₁ + y₂ [SMOD U] := by
+  rw [SModEq.def] at hxy₁ hxy₂⊢
+  simp_rw [Quotient.mk_add, hxy₁, hxy₂]
+#align smodeq.add SModEq.add
+
+theorem smul (hxy : x ≡ y [SMOD U]) (c : R) : c • x ≡ c • y [SMOD U] := by
+  rw [SModEq.def] at hxy⊢
+  simp_rw [Quotient.mk_smul, hxy]
+#align smodeq.smul SModEq.smul
+
+theorem zero : x ≡ 0 [SMOD U] ↔ x ∈ U := by rw [SModEq.def, Submodule.Quotient.eq, sub_zero]
+#align smodeq.zero SModEq.zero
+
+set_option synthInstance.etaExperiment true in
```

Comment:
@jcommelin, are you happy with this? Or do you want to try the local deinstancing fix instead?

If we do keep this, then probably we want
```suggestion
set_option synthInstance.etaExperiment true in  -- porting note: workaround for lean4#2074
```
(leanprover/lean4#2074)